### PR TITLE
allowing annotations for page dependencies

### PR DESCRIPTION
**Pull request by NikkyAI**
also other random fixes

### CODE REVIEW

File: core/src/main/kotlin/knote/PageManagerImpl.kt [hunk 1]
```diff
@@ -66,12 +60,36 @@ internal class PageManagerImpl(
...
+            logger.warn("script for page $pageId is not compiled")
+            evalPage(pageId)?.compiledScript ?: return null
+        }
+        val processFunction =
+            pageScript::class.declaredMemberFunctions.find { it.name == "process" } ?: run {
+                logger.error("no function `process` found in $pageId")
+                return null
+            }
+        return processFunction.returnType
+    }
```
Comment (ahinchman1) [inline core/src/main/kotlin/knote/PageManagerImpl.kt:78]: oh thank god we needed this for sure. I'll have to update this on the GUI accordingly when the merge goes up

File: core/src/main/kotlin/knote/PageManagerImpl.kt [hunk 2]
```diff
@@ -66,12 +60,36 @@ internal class PageManagerImpl(
...
+        }
+    }
+
+    override fun evalPage(pageId: String): Page? {
         val file = notebookScript.fileForPage(pageId, logger) ?: return null
         return evalPage(file, pageId)
     }
 
-    fun evalPage(file: File, id: String = file.name.substringBeforeLast(".page.kts")): Pair<PageScript, Any?>? {
+    fun evalPage(file: File, id: String = file.name.substringBeforeLast(".page.kts")): Page? {
```
Comment (ahinchman1) [inline core/src/main/kotlin/knote/PageManagerImpl.kt:92]: I'm glad you simplified the return here too

File: core/src/main/kotlin/knote/PageManagerImpl.kt [hunk 3]
```diff
@@ -88,16 +106,34 @@ internal class PageManagerImpl(
...
+            reports.forEach { report ->
+                val path = report.sourcePath?.let { "$it: " } ?: ""
+                val location = report.location?.posToString()?.let { "$it: " } ?: ""
+                val messageString = "$path$location${report.message}"
+                when (report.severity) {
+                    ScriptDiagnostic.Severity.FATAL -> logger.error { "FATAL: $messageString" }
+                    ScriptDiagnostic.Severity.ERROR -> logger.error { messageString }
+                    ScriptDiagnostic.Severity.WARNING -> logger.warn { messageString }
+                    ScriptDiagnostic.Severity.INFO -> logger.info { messageString }
+                    ScriptDiagnostic.Severity.DEBUG -> logger.debug { messageString }
```
Comment (ahinchman1) [inline core/src/main/kotlin/knote/PageManagerImpl.kt:118]: wow, error messaging too nice

File: core/src/main/kotlin/knote/api/PageManager.kt [hunk 4]
```diff
@@ -2,62 +2,16 @@ package knote.api
 
 import knote.script.PageScript
 import knote.util.KObservableMap
+import kotlin.reflect.KType
 import kotlin.script.experimental.api.ScriptDiagnostic
 
 interface PageManager {
     val pages: KObservableMap<String, Page>
```
Comment (ahinchman1) [inline core/src/main/kotlin/knote/api/PageManager.kt:9]: oh wait what are we doing with all that stuff then? I'm using PageManager rn, I'm gonna need to adjust this in the GUI too

File: core/src/main/kotlin/knote/script/PageConfiguration.kt [hunk 5]
```diff
@@ -39,61 +55,136 @@ class PageConfiguration : ScriptCompilationConfiguration({
...
+                        }
+                        depId to resultType
+                    }
+
+                logger.debug("generating dependencies for $notebookId:$pageId")
+                val file = PageDependency.generate(
+                    output = generatedSrc,
+                    notebookId = notebookId,
+                    pageId = pageId,
+                    pageDependencies = pageDependencies
```
Comment (ahinchman1) [inline core/src/main/kotlin/knote/script/PageConfiguration.kt:164]: Note for me: PageDependency is a file and the attributes that make it up

File: gradle-plugin/src/main/kotlin/knote/plugin/KNotePlugin.kt [hunk 6]
```diff
@@ -70,27 +72,95 @@ open class KNotePlugin : Plugin<Project> {
...
+                sourceDirs.add(notebookDir)
+                notebookDir
+                    .listFiles { _, name -> name.endsWith(".notebook.kts") }
+                    .forEach { scriptFile ->
+                        val id = scriptFile.name.substringBeforeLast(".notebook.kts")
+
+                        val generatedSrc = project.rootDir.resolve("build").resolve(".knote").resolve(id)
+                        val pagesSrc = project.rootDir.resolve("${id}_pages")
+
+                        sourceDirs.add(pagesSrc)
```
Comment (ahinchman1) [inline gradle-plugin/src/main/kotlin/knote/plugin/KNotePlugin.kt]: are these for the maven repository?

File: sample/msleep_pages/rangeSelection.page.kts [hunk 7]
```diff
@@ -1,6 +1,6 @@
-import krangl.DataFrame
+//@file:FromPage("sleepData")
 import krangl.*
 
-val sleepData: DataFrame = fromPage("sleepData")
+val sleepData: DataFrmae by inject()
```
Comment (ahinchman1) [inline sample/msleep_pages/rangeSelection.page.kts]: Mispelled `DataFrame` here

File: sample/statistics_pages/sequence.page.kts [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
-        Pair("Beta", 6.0),
-        Pair("Gamma", 7.2),
-        Pair("Delta", 9.2),
-        Pair("Epsilon", 6.8),
-        Pair("Zeta", 2.4),
-        Pair("Iota", 8.8)
+        Item("Alpha", 4.0),
+        Item("Beta", 6.0),
+        Item("Gamma", 7.2),
+        Item("Delta", 9.2),
```
Comment (ahinchman1) [inline sample/statistics_pages/sequence.page.kts:8]: hahahaha.... maybe rxJava would have been helpful after all

**Review by ahinchman1 — APPROVED** [review]
Some discussion items. Nothing to change.

File: core/src/main/kotlin/knote/PageManagerImpl.kt (hunk 2 above)
Comment (NikkyAI) [reply core/src/main/kotlin/knote/PageManagerImpl.kt:92]: this is somewhat still not what you are supposed to call from the GUI, there are better functions for also evaluating the page if necessary

File: core/src/main/kotlin/knote/PageManagerImpl.kt (hunk 3 above)
Comment (NikkyAI) [reply core/src/main/kotlin/knote/PageManagerImpl.kt:118]: the GUI can have a similar switch for colors

File: core/src/main/kotlin/knote/script/PageConfiguration.kt (hunk 5 above)
Comment (NikkyAI) [reply core/src/main/kotlin/knote/script/PageConfiguration.kt:164]: thats separate from the GUI, purely for doing the dependency stuff, the GUI should not use it, i think ?

File: gradle-plugin/src/main/kotlin/knote/plugin/KNotePlugin.kt (hunk 6 above)
Comment (NikkyAI) [reply gradle-plugin/src/main/kotlin/knote/plugin/KNotePlugin.kt]: this was my attempt to egister these folders as sourceFOlders in idea.. but it has no effect.. it seems to work without it thanks to a different feature i found, at least in my testing on the windows version of idea

File: sample/msleep_pages/rangeSelection.page.kts (hunk 7 above)
Comment (NikkyAI) [reply sample/msleep_pages/rangeSelection.page.kts]: that would explain it crashing on me.. oops

File: core/src/main/kotlin/knote/PageManagerImpl.kt (hunk 2 above)
Comment (ahinchman1) [reply core/src/main/kotlin/knote/PageManagerImpl.kt:92]: oh whoops I can change accordingly, I think I was doing `getResultOrExec`

**Pull request merged**